platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/DefaultValueSourceProviderFactory.java

```diff
@@ -278,33 +286,31 @@ private LazilyObtainedValue(
             this.sourceType = sourceType;
             this.parametersType = parametersType;
             this.parameters = parameters;
+            this.value = calculatedValueFactory.create(Describables.of("ValueSource of type", sourceType), () -> {
+                    computationBroadcaster.getSource().beforeValueObtained();
+                    try {
+                        ValueSource<T, P> source = source();
+                        sourceRef.set(source);
+                        return source.obtain();
+                    } finally {
+                        computationBroadcaster.getSource().afterValueObtained();
+                    }
+                }
+            );
         }
 
         public boolean hasBeenObtained() {
-            return value != null;
+            return value.isFinalized();
         }
 
-        public Try<T> obtain() {
-            ValueSource<T, P> source;
-            // Return value from local to avoid nullability warnings when returning value from the field directly.
-            Try<T> obtained;
-            synchronized (this) {
-                Try<T> cached = value;
-                if (cached != null) {
-                    return cached;
-                }
-                computationBroadcaster.getSource().beforeValueObtained();
-                try {
-                    // TODO - add more information to exceptions
-                    // Fail fast when source can't be instantiated.
-                    source = source();
```

Comment:
💭  There's a minor behavioral difference. Now the failure to create a ValueSource instance or isolate it parameters is cached too.